pkg/agent/multicast/mcast_discovery.go

```diff
@@ -97,14 +105,64 @@ func (s *IGMPSnooper) queryIGMP(group net.IP, versions []uint8) error {
 		if err != nil {
 			return err
 		}
-		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, mcastAllHosts, openflow13.P_NORMAL, igmp); err != nil {
+		outPort := uint32(0)
+		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, types.McastAllHosts, outPort, igmp); err != nil {
 			return err
 		}
-		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version)
+		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version, "outPort", outPort)
 	}
 	return nil
 }
 
+func (s *IGMPSnooper) validate(event *mcastGroupEvent) (bool, error) {
+	if s.validator == nil {
+		// Do not need to validate if the packet should be dropped since Validator is null, thus here returns directly
+		return true, nil
+	}
+	if event.iface.Type == interfacestore.ContainerInterface {
+		item, err := s.validator.Validate(event.iface.PodName, event.iface.PodNamespace, event.group)
+		klog.V(2).InfoS("call function s.Validator.Validate", "podName", event.iface.PodName,
+			"podNamespace", event.iface.PodNamespace, "mgroup", event.group.String())
+		if err != nil {
+			// Here just checks if packet should be dropped or not, Validate will return error when it fails to call ByIndex.
+			// We just imagine there is no rule for the pod(event.iface.PodName, event.iface.PodNamespace) to groupAddress event.group.
+			klog.ErrorS(err, "s.Validator.Validate returned")
+			return true, err
+		}
+		if item.NPType != nil {
+			klog.V(4).InfoS("s.Validator.Validate returned value", "RuleAction", *item.RuleAction,
+				"UUID", item.UUID, "Name", item.Name)
+			s.addToIGMPReportNPStatsMap(item, event.packetLen)
+		}
+		if item.RuleAction != nil && *item.RuleAction == v1alpha1.RuleActionDrop {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (s *IGMPSnooper) addToIGMPReportNPStatsMap(item types.McastNPValidationItem, packetLen uint64) {
```

Comment:
Question - when we delete stale entries from the map? I did not see the code for that.

Reply:
I will change the implementation later. Basically use
```
func (m *Collector) report(curStatsCollection *statsCollection) error {
```
to report igmpReportNPStatsMap every time and reset it. Will reply this thread after this update

Reply:
updated